### ISSUE TITLE
Qwen3.5-9B browser RL - is the language-only spec note still accurate, and are there reproduced numbers?

### ISSUE THREAD

**Issue by yerinNam** [OpenWebRL/OpenWebRL#4]
Hi, and thanks for open-sourcing this work.

I'm looking at `examples/orchard_gui/scripts/run_browser_qwen3.5_9b.sh` and have two questions before committing compute to a run.

**1. Vision spec**

The script header flags that `scripts/models/qwen3.5-9B.sh` is a language-only spec, with the note that this "builds a text-only model that cannot consume screenshot pixel_values" and that the agent would be "effectively blind" unless the browser task is driven by text observations.

Since the README now lists "Support RL with Qwen3.5 (2026.08.04)" as done, is that note stale? Concretely:

- Does the Megatron side train the vision tower, or is it frozen / excluded during the `--megatron-to-hf-mode bridge` conversion?
- If this config is intended to run on text observations only, it would help to say so explicitly in the header.

My understanding is that `Qwen/Qwen3.5-9B` is natively multimodal, so the rollout side (SGLang) should see screenshots regardless — the question is specifically about the training graph.

**2. Reproduced results**

Are there any success-rate numbers for the Qwen3.5-9B RL run (WebVoyager / Online-Mind2Web / DeepShop) that can be compared against the OpenWebRL-4B and -8B numbers in the paper?

The hyperparameters in the 9B script look inherited from the 8B recipe (lr 5e-7, ppo-epochs 2, eps-clip 0.2/0.28), so I'd like to know whether they've been validated at 9B or are a starting point.

Thanks!

**Comment by YangRui2015** [comment]
Hi, thank you for your interest in our work!

Qwen3.5-9B is indeed a multimodal model. The comment suggesting otherwise was mistakenly generated by Claude Code.

Yes, you can use OpenWebRL-9B-SFT (https://huggingface.co/OpenWebRL/OpenWebRL-9B-SFT) as the initialization checkpoint, as it achieves stronger performance than our 4B and 8B SFT models in our evaluation.

Regarding the 9B RL model, our progress has been slowed by limited compute resources and we will share an update when the results are ready. Thank you for your patience!

**Comment by yerinNam** [comment]
Thanks for the quick reply, and good to know the vision comment was spurious.

One follow-up on where the 9B RL path lives. This repo currently ships `scripts/run_browser_Qwen3VL_4B_Instruct.sh` and `..._8B_Instruct.sh` plus `train.py` and the vendored `slime/`, but the Qwen3.5 entry point in the README TODO points to `MSR-Orchard/slime` (`examples/orchard_gui/scripts/run_browser_qwen3.5_9b.sh`).

So, to make sure I set this up correctly:

1. Is the intended path for a 9B run to use the `MSR-Orchard/slime` fork directly, or will an equivalent `run_browser_qwen3.5_9b.sh` land in this repo alongside the 4B/8B scripts?
2. Does `openwebrl/` (env, reward, judge, response_format) get imported as-is from the slime side, or is there a separate copy there that may drift?

Happy to work off the slime fork if that's the supported route - just want to avoid mixing two versions of the harness. Thanks!

**Comment by QianhuiWu** [comment]
Hi @yerinNam, since the 9B run requires some newer dependencies that aren’t currently supported in the OpenWebRL/slime setup, we recommend running the 9B script from the Orchard project. Thanks!

**Comment by yerinNam** [comment]
Thanks @QianhuiWu !

I'll work off the Orchard project for the 9B run.
Two things from setting this up that seem worth reporting back.

**1. Vision tower in the training graph**
I think the training-graph half of my first question is still open. From the code:
- `slime_plugins/models/qwen3_5.py:get_qwen3_5_spec` builds a spec with no vision tower.
- `--megatron-to-hf-mode bridge` exports only the parameters actually present in the Megatron model.
If I'm reading that correctly, the ViT is excluded from the training graph and stays frozen at whatever weights SGLang loaded initially, while the rollout side uses the full HF VLM and does see screenshots normally. That would make the header's "effectively blind" claim wrong, but "language-only spec" accurate — the model trains as text-only even though it observes as multimodal.
The 4B script does the same thing (Qwen3-VL-4B-Thinking paired with the `qwen3-4B` spec), so a frozen ViT looks like the intended pattern across these recipes rather than a 9B-specific oversight. Could you confirm that's deliberate? If so, a one-line note in the script headers stating the ViT is not trained would be clearer than the current wording.

**2. Patch application (partly answers my Q2)**
`slime_f27ef35c.patch` doesn't apply against the vendored `slime/` in this repo in either direction, which blocks `--enable-adaptive-query-sampling`. Reading `setup_env.sh`, it looks like the intended flow is to clone upstream `f27ef35c` fresh and patch that, rather than patching the vendored tree — if that's right, it does confirm the two harnesses diverge. A short note on the expected directory layout for a 9B run would help people avoid mixing the two.
Happy to be corrected on either - I may be misreading the bridge path.

**Comment by QianhuiWu** [comment]
Hi @yerinNam, thanks for digging into this!

**1. Vision tower in the training graph**

Your reading would be exactly right in direct mode. But this recipe runs in bridge mode, and in bridge mode the spec is never used to build the model, so the premise "the training model comes from get_qwen3_5_spec" doesn't hold. Bridge mode bypasses the spec for model construction. In slime/backends/megatron_utils/model_provider.py:87, when megatron_to_hf_mode == "bridge", the provider comes from AutoBridge.from_hf_pretrained(args.hf_checkpoint).to_megatron_provider(), and the function returns before the args.spec branch is ever reached. get_qwen3_5_spec / scripts/models/qwen3.5-9B.sh (and the language-only converter megatron_to_hf/qwen3_5.py) only matter for non-bridge runs — the sourced MODEL_ARGS are still on the command line, but the model structure comes from the HF config.

And AutoBridge builds the full VLM for this checkpoint. Qwen/Qwen3.5-9B has architectures: ["Qwen3_5ForConditionalGeneration"], for which megatron-bridge registers Qwen35VLBridge with target=Qwen3VLModel (megatron/bridge/models/qwen_vl/qwen35_vl_bridge.py:267), including full ViT weight mappings (vision_model.* ↔ model.visual.*). You can confirm without a training run:
```bash
>>> bridge = AutoBridge.from_hf_pretrained("Qwen/Qwen3.5-9B", trust_remote_code=True)
>>> type(bridge.model_bridge).__name__
'Qwen35VLBridge'
>>> type(bridge.to_megatron_provider(load_weights=False)).__name__
'Qwen35VLModelProvider'   # freeze_vision_model defaults to False
```

So your second premise is true but cuts the other way: bridge export does only cover params present in the Megatron model — and vision_model.* is present. It's loaded from HF at start (checkpoint.py:141, bridge.load_hf_weights), it's in the training graph (rollout screenshots reach the training forward as pixel_values via multimodal_train_inputs, model.py:431), nothing freezes it (provider freeze flags default to False, and no recipe sets --freeze-params-name-list), and it's exported on every weight sync. The ViT trains.

The 4B/8B Qwen3-VL recipes follow the same pattern rather than a frozen-ViT-by-design: they also set --megatron-to-hf-mode bridge, so the qwen3-4B spec they source is ignored for construction in the same way.

That said, you're right that the header was misleading, and the direct-mode path really is language-only — switching off bridge mode would silently drop the ViT from training, sync, and --save-hf exports. We'll update the 9B script wording accordingly, including a warning against exactly that switch.

**2. Patch application**

We cloned upstream f27ef35c fresh and patch that, which does not depend on OpenWebRL. Please follow the deployment instructions directly in the [Project Orchard README under slime/examples/orchard_gui](https://github.com/MSR-Orchard/slime/tree/main/examples/orchard_gui). [official slime docker v0.3.0]((https://hub.docker.com/r/slimerl/slime)) should work well with this setup.

Thanks again for flagging these — we’ll make the documentation clearer to avoid confusion between the two setups.
